### ISSUE TITLE
Implement JWT Authorization Grants

### ISSUE THREAD

**Issue by rlewczuk** [spring-projects/spring-authorization-server#546]
Implement JWT Authorization Grant as defined in RFC 7523.

**Comment by fkhantsi** [comment]
For those inquiring about the future of this feature, @jgrandja has [confirmed](https://github.com/spring-projects/spring-authorization-server/issues/966#issuecomment-1324267304) its scheduled for release `1.1.0`

**Comment by jgrandja** [comment]
@fkhantsi Correction. This feature is not scheduled as of yet. The link states:

> We are also planning on documenting "Adding an extension authorization grant" in the reference manual via https://github.com/spring-projects/spring-authorization-server/issues/686. I scheduled it for the 1.1.0 release.

To clarify, gh-686 is scheduled for `1.1.0`.

**Comment by imwangyi** [comment]
is there any sample for using JWT Bearer? I noticed 1.1.0 had been released.

**Comment by jgrandja** [comment]
@bhashitbhatt There are no plans as of yet.

We have a number of higher priority items on our list for the major release of Spring Security `7.0` and Spring Authorization Server `2.0` and with less resources on our team it will limit the new features in the next release.

**Comment by symposion** [comment]
> Refresh tokens should not be returned.

Yeah, I was afraid you'd say this :-/  I agree with you, but unfortunately Google have taken a different view on this. They're using this grant as part of an account linking process, and because the JWT transmitted as part of it contains a limited amount of Personally Identifying Data, they only want to send it while the user is interactively engaged with the process; but they still want to be able to then use the linked account for offline access later on - thus the requirement for refresh tokens. 

I completely understand that you probably don't want to build this in to the basic implementation - I agree that the intention of the spec is clear here, even if it does stop short of explicitly prohibiting the return of a refresh token (it only says "that will usually mean...."); but given that Google is Google and this is a pretty significant use case for this grant, I think it would be great if we can find a way to implement it that leaves open the possibility of returning a refresh token by some kind of relatively limited customisation. Not to put too fine a point on it, we absolutely have to have this functionality as a result of Google's choices here, so we will be implementing it in a way that permits this whatever happens. The only question is whether we can find a way to do it that lets us contribute our implementation back to the project, or if it has to be a completely bespoke implementation. I think it would be a shame if we ended up there, particularly as this is standard part of Google's IAM mechanisms, so I'm sure there will be others that need to do the same.

**Comment by symposion** [comment]
> I believe the implementation would be very similar to OAuth2ClientCredentialsAuthenticationProvider

One interesting problem/question here is that you can - and frequently will - have both an authenticated client _and_ (sort of) an authenticated end user (as specified by the JWT assertion). I'm assuming that we will need to reserve the main security context for any possible client authentication (as happens for e.g. the client_credentials grant) and then handle the JWT assertion of the end user locally to the authentication provider for this specific grant. 

However, the closest parallel that I can find for validating the JWT assertion would be Spring Security's  implementation of OAuth2 login, which is very much tied to creating an interactive login session. I suspect that trying to leverage much of this - despite the close parallels - is likely to just result in more confusion and maintenance issues. Also potentially relevant is the IDP federation sample, which is dealing with some of the same problems (handling an external IDP assertion of identity).

It might be a bit ambitious but it would be fantastic if we could find a way to align the handling of all of these at least to some extent. The basic problem is more or less the same in each case:

 * External assertion of identity in the form of a JWT with some kind of identifiers
 * Process that assertion - possibly by looking up against an existing user store (e.g. OidcUserService)
 * Potentially handle on-the-fly local user creation if appropriate
 * Produce access token (and maybe other tokens) based on the resulting user (for JWT Auth Grant/Federated IDP) OR add them to the security context as the logged in user (for OAuth2 Login)

It's really the middle of that process that's common to all of these - translating an external JWT to a local user - and I think it would be a great outcome if we could find a way to achieve that without loads of code duplication/completely separate mechanisms
